src/Datadog.Trace/TracingProcessManager.cs

```diff
@@ -115,65 +123,112 @@ private static Task StartProcessWithKeepAlive(ProcessMetadata metadata)
                             {
-                                metadata.IsFaulted = false;
+                                Log.Error("Circuit breaker triggered for {Process}. Max failed retries reached ({ErrorCount}).", path, MaxFailures);
+                                metadata.ProcessState = ProcessState.Faulted;
+                                return;
+                            }
 
-                                if (metadata.ProgramIsRunning())
+                            try
+                            {
+                                if (metadata.ProcessState == ProcessState.NotRunning)
                                 {
-                                    Log.Debug("{Process} is already running.", path);
-                                    continue;
+                                    // This means we have never tried from this domain
+                                    var pipeIsBound = metadata.NamedPipeIsBound();
+
+                                    if (pipeIsBound)
+                                    {
+                                        // It is possible that if a pipe is bound it may still be shutting down from last time
+                                        // Check on a delay to be sure we have the agent available
+                                        var attempts = 7;
+                                        var delay = 50d;
+
+                                        while (--attempts > 0)
+                                        {
+                                            await Task.Delay((int)delay);
+                                            if (!metadata.NamedPipeIsBound())
+                                            {
+                                                metadata.ProcessState = ProcessState.ReadyToStart;
+                                                break;
+                                            }
+
+                                            // Should result in a max delay of ~3.28 seconds before giving up
+                                            delay = delay * 1.75d;
+                                        }
+
+                                        metadata.ProcessState = ProcessState.Healthy;
+                                    }
+                                    else
+                                    {
+                                        // If no pipe is bound, kick it off
+                                        metadata.ProcessState = ProcessState.ReadyToStart;
+                                    }
                                 }
+                                else if (metadata.ProcessState == ProcessState.Healthy || metadata.ProcessState == ProcessState.Faulted)
+                                {
+                                    // This means we have tried to start from this domain before and we're in a keep alive check
+                                    if (!metadata.NamedPipeIsBound())
```

Comment:
nit: I'd prefer to not have negation in the condition where possible, so `if (metadata.NamedPipeIsBound())`